enhancements/cluster-logging/loki-storage.md

```diff
@@ -0,0 +1,628 @@
+    // are degraded or the cluster cannot connect to object storage.
+    ConditionDegraded LokiStackConditionType = "Degraded"
+)
+
+// LokiStackConditionReason defines the type for valid reasons of a Loki deployment conditions.
+type LokiStackConditionReason string
+
+const (
+    // ReasonMissingObjectStorageSecret when the required secret to store logs to object
+    // storage is missing.
+    ReasonMissingObjectStorageSecret LokiStackConditionReason = "MissingObjectStorageSecret"
+
+    // ReasonInvalidObjectStorageSecret when the format of the secret is invalid.
+    ReasonInvalidObjectStorageSecret LokiStackConditionReason = "InvalidObjectStorageSecret"
+
+    // ReasonInvalidRedundancyConfiguration when the configurated redudancy is not valid
+    // with the select cluster size.
+    ReasonInvalidRedundancyConfiguration LokiStackConditionReason = "InvalidRedundancyConfiguration"
+)
+
+// LokiStackStatus defines the observed state of LokiStack
+type LokiStackStatus struct {
+    // Conditions of the Loki deployment health.
+    //
+    // +optional
+    Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+```
+
+### clusterlogging_types.go
+
+The integration of the `LokiStack` CRD into `ClusterLogging/v1` custom resources is following the present pattern to first into the latter.
+
+```go
+type LogStoreType string
+
+const (
+    LogStoreTypeElasticsearch LogStoreType = "elasticsearch"
+
+    LogStoreTypeLoki  LogStoreType = "Loki"
+)
+
+type LogStoreSpec struct {
+    // The type of Log Storage to configure
+    Type LogStoreType `json:"type`"
+
+    LokiStackSpec `json:"Loki,omitempty"`
+}
+```
+
+### Test Plan
+
+#### Unit Testing
+
+The [loki-operator](https://github.com/viaq/loki-operator) includes a framework based on [counterfeiter](github.com/maxbrunsfeld/counterfeiter) to create simple and useful fake client to test configuration in unit tests.
+
+Testing of the reconciliation of a `LokiStack` custom resource will be based upon the same technique where the custom resource describes the inputs and the generated manifests the outputs.
+
+#### Functional Testing
+
+The [cluster-logging-operator](https://github.com/openshift/cluster-logging-operator) provides a simple [functional test framework](https://github.com/openshift/cluster-logging-operator/tree/master/test/functional) to test the configuration provided in a `ClusterLogging` custom resource to collector configuration.
+
+Testing the integration of the `LokiStack` custom resource in the `ClusterLogging` custom resource will be based upon this technique.
+
+#### E2E Testing
+
+Finally E2E testing requiring a running cluster in form of [kind](https://kind.sigs.k8s.io/) or [OpenShift](https://github.com/openshift/) is dedicated later for:
+
+1. Upgrade `LokiStack` scenarios.
+2. Compatibility testing with various S3-compatible object-storage providers (e.g. Openshift Storage).
+3. Integration with various caching providers, e.g. Memcached, Redis.
```

Comment:
This is a bit sudden. Where do these fit into the picture, how are they deployed, controlled etc.?

Reply:
These are optional performance-oriented dependencies for query performance. The path forward here is to support them via their dedicated operators, i.e. loki-operator will create eventually a Memcached CR or a Redis CR.

Reply:
I disagree. I don't think Loki will create any CR. We should ask the user for endpoints to an existing system. If we want to enable caching then that would live in CLO. Loki Operator is not an operator of operators it's just an operator of Loki.

Reply:
I too think operator composition needs to be left to a higher-level operator, like CLO.

Reply:
Agreed! We can keep composition the ClusterLogging API and to cluster-logging-operator to drive the CRs for caching, etc.

Besides that this subsection discusses the E2E testing use cases for Loki. One of them is to test Loki with caching. Should we mention that we need CLO for such an E2E test? I believe not because:
- Testing Loki with caching requires available caching instance only not CLO creating caching CRs.
- Limiting the amount of dependencies (e.g. operators) one needs to run the tests is by experience a good path toward limiting flakes to happen.

@syedriko @blockloop WDYT? How should this sentence be improved? Mentioning CLO explicitly as a dependency for testing or not would help the reader?

Reply:
@periklis I would keep testing of caching to CLO